lib/platform/linux-uart.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <signal.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+static const char *dbginfo_off = "pscdbg 0 all\r";
+static const char *echo_off = "echo 0\r";
+static const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(10000)
+
+enum switchtec_uart_cmd_type {
+	SWITCHTEC_UART_CONTROL = 0,
+	SWITCHTEC_UART_GAS_READ = 1,
+	SWITCHTEC_UART_GAS_WRITE = 2,
+};
+
+static int send_cmd(int fd, const char *str, enum switchtec_uart_cmd_type type,
+		    uint32_t addr, size_t n, const void *src, uint8_t crc)
+{
```

Comment:
Try something like this. Having special cases in a function like this is not very good.

```c
#include <stdio.h>
#include <unistd.h>
#include <stdint.h>
#include <errno.h>

#include <stdarg.h>

/**
 * @brief Send a formatted command out to the UART device
 * @param[in] fd			File descriptor to send the command to
 * @param[in] fmt			Format string passed to sprintf
 * @param[in] write_bytes   Number of bytes in an array to append to the string
 * @param[in] *format_args  Arguments passed to sprintf
 * @param[in] write_data    If write_bytes is non-zero, a pointer to
 *                          data of size \p write_bytes to append to
 *                          the string
 * @param[in] crc           If write_bytes is non-zero, an extra CRC byte
 *                          to append to the data
 * @return 0 on success, negative on error
 */
static int send_cmd(int fd, const char *fmt, int write_bytes, ...)
{
    char cmd[1024];
    int cnt;
    uint8_t *write_data;
    int i;
    int ret;
    va_list argp;

    va_start(argp, write_bytes);

    cnt = vsnprintf(cmd, sizeof(cmd), fmt, argp);

    if (write_bytes) {
        write_data = va_arg(argp, uint8_t *);
        for (i = 0; i < write_bytes; i++) {
            cnt += snprintf(cmd + cnt, sizeof(cmd) - cnt,
                            "%02x", write_data[i]);
        }

        //Extra arg for the CRC
        cnt += snprintf(cmd + cnt, sizeof(cmd) - cnt,
                        "%02x\n", va_arg(argp, int));

    }

    va_end(argp);

    ret = write(fd, cmd, cnt);
    if (ret < 0)
        return ret;

    if (ret != cnt) {
        errno = EIO;
        return -errno;
    }

    return 0;
}

void main(void)
{
    unsigned long addr = 0xAABBCCDD;
    uint8_t data[] = {0xEE, 0xFF, 0x11, 0x22, 0x33};
    uint8_t crc = 0x44;

    send_cmd(STDOUT_FILENO, "control command\n", 0);
    send_cmd(STDOUT_FILENO, "gaswr -c -s 0x%x 0x", sizeof(data),
             addr, data, crc);
    send_cmd(STDOUT_FILENO, "gasrd -c -s 0x%x %zu\n", 0, addr, 5);
}

```

Reply:
change to variable parameters function. thx.